src/utils/string.rs

```diff
@@ -17,98 +17,88 @@ pub(crate) fn string_from_lfn(utf16_src: &[u16]) -> Result<String, FromUtf16Erro
     String::from_utf16(&utf16_src[..nul_range_end])
 }
 
-pub(crate) fn as_sfn(string: &str, codepage: &Codepage) -> Option<Sfn> {
-    // anything non-ascii should be represented as a LFN
-    if !string.chars().all(|c| codepage.contains(c)) {
-        return None;
-    }
+pub(crate) fn as_sfn(string: &str, codepage: Codepage) -> Option<Sfn> {
+    // a file can still not have an extension
+    let (name, ext) = string.split_once('.').unwrap_or((string, ""));
 
-    // since everything is in ascii, 1 byte = 1 char
-    if string.len() > SFN_NAME_LEN + 1 + SFN_EXT_LEN {
-        return None;
-    }
+    // create a sfn with padding
+    let mut result = Sfn {
+        name: *b"        ",
+        ext: *b"   ",
+    };
 
-    // all alphabetic characters must be capitalized
-    if !string
-        .chars()
-        .filter(|c| c.is_alphabetic())
-        .all(|c| c.is_uppercase())
-    {
-        return None;
-    }
+    copy_cp_chars(&mut result.name, name, codepage)?;
 
-    // a file can still not have an extension
-    let (name, ext) = string.split_once('.').unwrap_or((string, ""));
+    copy_cp_chars(&mut result.ext, ext, codepage)?;
+
+    Some(result)
+}
 
-    if name.len() > SFN_NAME_LEN || ext.len() > SFN_EXT_LEN {
-        return None;
+fn copy_cp_chars(mut destination: &mut [u8], string: &str, codepage: Codepage) -> Option<()> {
+    for c in string.chars() {
+        let c = encode_valid_char_checked(c, codepage)?;
+        if destination.is_empty() {
+            // no space left
+            return None;
+        }
+        destination[0] = c;
+        destination = &mut destination[1..];
     }
 
-    // don't forget the padding
-    // here, 8 is SFN_NAME_LEN and 3 is SFN_EXT_LEN
-    let (name, ext) = (format!("{name:<8}"), format!("{ext:<3}"));
+    Some(())
+}
 
-    // this shouldn't panic, as we have checked that both the name
-    // and the extension should fit into a SFN
-    Some(Sfn {
-        name: name.as_bytes().try_into().unwrap(),
-        ext: ext.as_bytes().try_into().unwrap(),
-    })
+fn encode_valid_char_checked(c: char, codepage: Codepage) -> Option<u8> {
+    let c = codepage.encode_char_checked(c)?;
+    if c.is_ascii_digit() || c.is_ascii_uppercase() || b"$%-_@~`!(){}^#&".contains(&c) || c >= 128 {
+        return Some(c);
+    }
+    None
```

Comment:
The purpose of these functions isn't that obvious to me. Could you please elaborate?

Reply:
the `copy_cp_chars` fn is called for both name and extension.

since the size is no longer checked ahead of time (which was not working anyway, since it assumed it's ascii but allowed other cp* char which are utf8 encoded) now it needs to check if it fits into the 8/3 byte space.

also if the char is not valid it's rejected (valid are 0-9A-Z some more ascii chars and all non-ascii-chars, according to some websites).

Reply:
> also if the char is not valid it's rejected (valid are 0-9A-Z some more ascii chars and all non-ascii-chars, according to some websites).

Yeah, I forgot to check that. I am gonna switch to using `typed_path::constants::windows::DISALLOWED_FILENAME_BYTES` or `DISALLOWED_FILENAME_CHARS` tho, since it does the same job an is more readable imo

Reply:
Alright, thanks